File: cases/06-broken-pipeline-and-fragile-delivery/java/app/Main.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.LongAdder;

/**
 * Caso 06 — Pipeline roto y entrega fragil (stack Java).
 *
 * Legacy: deploy directo sin preflight, sin smoke, sin rollback.
 * Controlled: state machine (preflight → canary → smoke → promote | rollback).
 *
 * Primitivas Java distintivas:
 *   - record types para DeploymentState inmutables.
 *   - ConcurrentHashMap<String,EnvState> para snapshot por ambiente sin lock global.
 *   - State machine como enum + transition table.
 */
public class Main {

    private static final String CASE_NAME = "06 - Pipeline roto y delivery fragil";
    private static final String STACK = "Java 21";
    private static final int PORT = Integer.parseInt(System.getenv().getOrDefault("PORT", "8080"));
    private static final int MAX_DEPLOYMENTS = 30;

    private static final Map<String, EnvState> environments = new ConcurrentHashMap<>();
    private static final Deque<Deployment> deployments = new ArrayDeque<>();

    private static final LongAdder legacyDeploys = new LongAdder();
    private static final LongAdder legacyBroken = new LongAdder();
    private static final LongAdder controlledDeploys = new LongAdder();
    private static final LongAdder controlledRollbacks = new LongAdder();
    private static final LongAdder controlledBlocked = new LongAdder();

    public static void main(String[] args) throws Exception {
        environments.put("staging", new EnvState("staging", "v1.0.0", "healthy"));
        environments.put("prod", new EnvState("prod", "v1.0.0", "healthy"));
        HttpServer server = HttpServer.create(new InetSocketAddress(PORT), 0);
        server.createContext("/", Main::route);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("[case06-java] listening on " + PORT);
        Runtime.getRuntime().addShutdownHook(new Thread(() -> server.stop(0)));
    }

    private static void route(HttpExchange ex) throws IOException {
        URI uri = ex.getRequestURI();
        String path = uri.getPath();
        Map<String, String> q = queryParams(uri.getRawQuery());
        int status = 200;
        String body;

        try {
            switch (path) {
                case "/":
                case "/index":
                    body = "{\"case\":\"" + CASE_NAME + "\",\"stack\":\"" + STACK +
                            "\",\"routes\":[\"/health\",\"/deploy-legacy?env=prod&version=v1.1.0&scenario=secret_drift\",\"/deploy-controlled?env=prod&version=v1.1.0&scenario=secret_drift\",\"/environments\",\"/deployments\",\"/diagnostics/summary\",\"/reset-lab\"]}";
                    break;
                case "/health":
                    body = "{\"status\":\"ok\",\"stack\":\"" + STACK + "\",\"case\":\"" + CASE_NAME + "\"}";
                    break;
                case "/deploy-legacy":
                    body = deployLegacy(q.getOrDefault("env", "prod"),
                            q.getOrDefault("version", "v1.1.0"),
                            q.getOrDefault("scenario", "clean"));
                    break;
                case "/deploy-controlled":
                    body = deployControlled(q.getOrDefault("env", "prod"),
                            q.getOrDefault("version", "v1.1.0"),
                            q.getOrDefault("scenario", "clean"));
                    break;
                case "/environments":
                    body = environmentsJson();
                    break;
                case "/deployments":
                    body = deploymentsJson();
                    break;
                case "/diagnostics/summary":
                    body = diagnosticsJson();
                    break;
                case "/reset-lab":
                    environments.put("staging", new EnvState("staging", "v1.0.0", "healthy"));
                    environments.put("prod", new EnvState("prod", "v1.0.0", "healthy"));
                    synchronized (deployments) { deployments.clear(); }
                    legacyDeploys.reset(); legacyBroken.reset();
                    controlledDeploys.reset(); controlledRollbacks.reset(); controlledBlocked.reset();
                    body = "{\"status\":\"reset\"}";
                    break;
                default:
                    status = 404;
                    body = "{\"error\":\"not_found\",\"path\":\"" + escape(path) + "\"}";
            }
        } catch (Exception e) {
            status = 500;
            body = "{\"error\":\"internal\",\"detail\":\"" + escape(e.getMessage()) + "\"}";
        }

        byte[] out = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().add("Content-Type", "application/json; charset=utf-8");
        ex.sendResponseHeaders(status, out.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(out); }
    }

    /** Legacy: aplica el version sin preflight, deja ambiente roto si el scenario falla. */
    private static String deployLegacy(String env, String version, String scenario) {
        legacyDeploys.increment();
        String result;
        String health;
        if (isBadScenario(scenario)) {
            health = "degraded";
            legacyBroken.increment();
            result = "deployed_but_broken";
        } else {
            health = "healthy";
            result = "deployed";
        }
        environments.put(env, new EnvState(env, version, health));
        record("legacy", env, version, scenario, result);
        return "{\"variant\":\"legacy\",\"env\":\"" + env + "\",\"version\":\"" + version +
                "\",\"scenario\":\"" + scenario + "\",\"result\":\"" + result + "\",\"health\":\"" + health +
                "\",\"note\":\"sin preflight ni rollback; ambiente queda como quede.\"}";
    }

    /** Controlled: preflight → smoke → promote, o rollback si smoke falla. */
    private static String deployControlled(String env, String version, String scenario) {
        controlledDeploys.increment();
        EnvState before = environments.get(env);
        // preflight bloquea si scenario es claramente malo
        if (scenario.equals("missing_artifact") || scenario.equals("secret_drift_detected")) {
            controlledBlocked.increment();
            record("controlled", env, version, scenario, "blocked_in_preflight");
            return "{\"variant\":\"controlled\",\"env\":\"" + env + "\",\"version\":\"" + version +
                    "\",\"scenario\":\"" + scenario + "\",\"result\":\"blocked_in_preflight\"," +
                    "\"current_version\":\"" + before.version + "\"," +
                    "\"note\":\"preflight bloqueo antes de tocar el ambiente.\"}";
        }
        // smoke despues del deploy
        if (isBadScenario(scenario)) {
            // rollback automatico
            controlledRollbacks.increment();
            record("controlled", env, version, scenario, "rolled_back_to_" + before.version);
            return "{\"variant\":\"controlled\",\"env\":\"" + env + "\",\"version\":\"" + version +
                    "\",\"scenario\":\"" + scenario + "\",\"result\":\"rolled_back\"," +
                    "\"current_version\":\"" + before.version + "\"," +
                    "\"note\":\"smoke fallo, rollback automatico al version anterior.\"}";
        }
        environments.put(env, new EnvState(env, version, "healthy"));
        record("controlled", env, version, scenario, "promoted");
        return "{\"variant\":\"controlled\",\"env\":\"" + env + "\",\"version\":\"" + version +
                "\",\"scenario\":\"" + scenario + "\",\"result\":\"promoted\",\"health\":\"healthy\"," +
                "\"note\":\"preflight ok + smoke ok → promote.\"}";
    }

    private static boolean isBadScenario(String scenario) {
        return scenario.equals("secret_drift") || scenario.equals("breaking_change")
                || scenario.equals("schema_mismatch");
    }

    private static void record(String variant, String env, String version, String scenario, String result) {
        synchronized (deployments) {
            deployments.addFirst(new Deployment(Instant.now().toString(), variant, env, version, scenario, result));
            while (deployments.size() > MAX_DEPLOYMENTS) deployments.removeLast();
        }
    }

    private static String environmentsJson() {
        StringBuilder sb = new StringBuilder(512);
        sb.append("{\"envs\":[");
        boolean first = true;
        for (EnvState s : environments.values()) {
            if (!first) sb.append(',');
            sb.append("{\"name\":\"").append(s.name).append("\",\"version\":\"").append(s.version)
              .append("\",\"health\":\"").append(s.health).append("\"}");
            first = false;
        }
        sb.append("]}");
        return sb.toString();
    }

    private static String deploymentsJson() {
        StringBuilder sb = new StringBuilder(1024);
        sb.append("{\"history\":[");
        synchronized (deployments) {
            boolean first = true;
            for (Deployment d : deployments) {
                if (!first) sb.append(',');
                sb.append("{\"at\":\"").append(d.at).append("\",\"variant\":\"").append(d.variant)
                  .append("\",\"env\":\"").append(d.env).append("\",\"version\":\"").append(d.version)
                  .append("\",\"scenario\":\"").append(d.scenario).append("\",\"result\":\"").append(d.result).append("\"}");
                first = false;
            }
        }
        sb.append("],\"max_kept\":").append(MAX_DEPLOYMENTS).append('}');
        return sb.toString();
    }

    private static String diagnosticsJson() {
        return "{\"stack\":\"" + STACK + "\",\"case\":\"" + CASE_NAME + "\"," +
                "\"legacy\":{\"deploys\":" + legacyDeploys.sum() +
                ",\"broken_state_left\":" + legacyBroken.sum() +
                ",\"behavior\":\"sin preflight, sin rollback\"}," +
                "\"controlled\":{\"deploys\":" + controlledDeploys.sum() +
                ",\"blocked_in_preflight\":" + controlledBlocked.sum() +
                ",\"rollbacks\":" + controlledRollbacks.sum() +
                ",\"behavior\":\"preflight + smoke + rollback automatico\"}," +
                "\"environments\":" + environmentsJson() + "}";
    }

    private record EnvState(String name, String version, String health) {}
    private record Deployment(String at, String variant, String env, String version, String scenario, String result) {}

    private static String escape(String v) {
        return v == null ? "" : v.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static Map<String, String> queryParams(String rawQuery) {
        Map<String, String> params = new HashMap<>();
        if (rawQuery == null || rawQuery.isBlank()) return params;
        for (String pair : rawQuery.split("&")) {
            String[] parts = pair.split("=", 2);
            String k = URLDecoder.decode(parts[0], StandardCharsets.UTF_8);
            String v = parts.length > 1 ? URLDecoder.decode(parts[1], StandardCharsets.UTF_8) : "";
            params.put(k, v);
        }
        return params;
    }
}

File: cases/06-broken-pipeline-and-fragile-delivery/rust/app/src/main.rs
// Caso 06 — Pipeline roto y entrega fragil (stack Rust 1.83).
//
// Legacy: deploy directo sin preflight, sin smoke, sin rollback.
// Controlled: preflight → deploy → smoke → promote | rollback.
//
// El contraste que este stack aporta:
//
//   Los resultados posibles del pipeline se modelan con un `enum` con datos
//   asociados, no con strings sueltos:
//
//       enum DeployOutcome {
//           Deployed,
//           DeployedButBroken,
//           BlockedInPreflight { current_version: String },
//           RolledBack { to_version: String },
//           Promoted,
//       }
//
//   Y el `match` sobre ese enum es **exhaustivo**: si mañana alguien agrega
//   una variante `Canary`, todos los `match` que no la contemplen dejan de
//   compilar. En Java, .NET, Go o Node el resultado es un string y agregar un
//   estado nuevo no rompe nada — simplemente cae al `else` de algun `if` y se
//   comporta como si fuera otra cosa.
//
//   Para una maquina de estados de deploy, esa diferencia no es estetica: el
//   estado no contemplado es exactamente el que deja produccion a medio camino.

use std::collections::HashMap;
use std::io::{BufRead, BufReader, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Mutex;
use std::thread;
use std::time::{SystemTime, UNIX_EPOCH};

const CASE_NAME: &str = "06 - Pipeline roto y delivery fragil";
const MAX_DEPLOYMENTS: usize = 30;

static LEGACY_DEPLOYS: AtomicI64 = AtomicI64::new(0);
static LEGACY_BROKEN: AtomicI64 = AtomicI64::new(0);
static CONTROLLED_DEPLOYS: AtomicI64 = AtomicI64::new(0);
static CONTROLLED_ROLLBACKS: AtomicI64 = AtomicI64::new(0);
static CONTROLLED_BLOCKED: AtomicI64 = AtomicI64::new(0);

#[derive(Clone)]
struct EnvState {
    name: String,
    version: String,
    health: String,
}

#[derive(Clone)]
struct Deployment {
    at: String,
    variant: String,
    env: String,
    version: String,
    scenario: String,
    result: String,
}

/// Resultados posibles del pipeline. El match sobre este enum es exhaustivo:
/// agregar una variante rompe la compilacion de quien no la contemple.
enum DeployOutcome {
    Deployed,
    DeployedButBroken,
    BlockedInPreflight { current_version: String },
    RolledBack { to_version: String },
    Promoted,
}

impl DeployOutcome {
    fn label(&self) -> String {
        match self {
            DeployOutcome::Deployed => "deployed".to_string(),
            DeployOutcome::DeployedButBroken => "deployed_but_broken".to_string(),
            DeployOutcome::BlockedInPreflight { .. } => "blocked_in_preflight".to_string(),
            DeployOutcome::RolledBack { to_version } => format!("rolled_back_to_{to_version}"),
            DeployOutcome::Promoted => "promoted".to_string(),
        }
    }
}

struct State {
    environs: Vec<EnvState>,
    deployments: Vec<Deployment>,
}

static STATE: Mutex<Option<State>> = Mutex::new(None);

fn stack() -> String {
    std::env::var("APP_STACK").unwrap_or_else(|_| "Rust 1.83".to_string())
}

fn fresh_state() -> State {
    State {
        environs: vec![
            EnvState { name: "staging".into(), version: "v1.0.0".into(), health: "healthy".into() },
            EnvState { name: "prod".into(), version: "v1.0.0".into(), health: "healthy".into() },
        ],
        deployments: Vec::new(),
    }
}

fn main() {
    *STATE.lock().unwrap() = Some(fresh_state());

    let port: u16 = std::env::var("PORT").ok().and_then(|p| p.parse().ok()).unwrap_or(8080);
    let listener = TcpListener::bind(("0.0.0.0", port)).expect("bind");
    println!("[case06-rust] listening on {port}");
    for stream in listener.incoming().flatten() {
        thread::spawn(move || handle_conn(stream));
    }
}

// ---------- capa HTTP minima ----------

fn handle_conn(mut stream: TcpStream) {
    let mut reader = BufReader::new(match stream.try_clone() {
        Ok(s) => s,
        Err(_) => return,
    });
    let mut request_line = String::new();
    if reader.read_line(&mut request_line).is_err() {
        return;
    }
    loop {
        let mut line = String::new();
        match reader.read_line(&mut line) {
            Ok(0) => break,
            Ok(_) if line.trim().is_empty() => break,
            Ok(_) => continue,
            Err(_) => break,
        }
    }
    let target = request_line.split_whitespace().nth(1).unwrap_or("/");
    let (path, query) = target.split_once('?').unwrap_or((target, ""));
    let params = parse_query(query);
    let (status, body) = route(path, &params);
    let reason = if status == 200 { "OK" } else { "Not Found" };
    let response = format!(
        "HTTP/1.1 {status} {reason}\r\nContent-Type: application/json; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}",
        body.as_bytes().len()
    );
    let _ = stream.write_all(response.as_bytes());
    let _ = stream.flush();
}

fn parse_query(raw: &str) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for pair in raw.split('&').filter(|p| !p.is_empty()) {
        let (k, v) = pair.split_once('=').unwrap_or((pair, ""));
        out.insert(url_decode(k), url_decode(v));
    }
    out
}

fn url_decode(s: &str) -> String {
    s.replace('+', " ")
}

// ---------- routing ----------

fn route(path: &str, params: &HashMap<String, String>) -> (u16, String) {
    let env = params.get("env").cloned().unwrap_or_else(|| "prod".into());
    let version = params.get("version").cloned().unwrap_or_else(|| "v1.1.0".into());
    let scenario = params.get("scenario").cloned().unwrap_or_else(|| "clean".into());

    match path {
        "/" | "/index" => (
            200,
            format!(
                r#"{{"case":"{CASE_NAME}","stack":"{}","routes":["/health","/deploy-legacy?env=prod&version=v1.1.0&scenario=secret_drift","/deploy-controlled?env=prod&version=v1.1.0&scenario=secret_drift","/environments","/deployments","/diagnostics/summary","/reset-lab"]}}"#,
                stack()
            ),
        ),
        "/health" => (
            200,
            format!(r#"{{"status":"ok","stack":"{}","case":"{CASE_NAME}"}}"#, stack()),
        ),
        "/deploy-legacy" => (200, deploy_legacy(&env, &version, &scenario)),
        "/deploy-controlled" => (200, deploy_controlled(&env, &version, &scenario)),
        "/environments" => (200, environments_json()),
        "/deployments" => (200, deployments_json()),
        "/diagnostics/summary" => (200, diagnostics()),
        "/reset-lab" => {
            *STATE.lock().unwrap() = Some(fresh_state());
            LEGACY_DEPLOYS.store(0, Ordering::Relaxed);
            LEGACY_BROKEN.store(0, Ordering::Relaxed);
            CONTROLLED_DEPLOYS.store(0, Ordering::Relaxed);
            CONTROLLED_ROLLBACKS.store(0, Ordering::Relaxed);
            CONTROLLED_BLOCKED.store(0, Ordering::Relaxed);
            (200, r#"{"status":"reset"}"#.to_string())
        }
        _ => (
            404,
            format!(r#"{{"error":"not_found","path":"{}"}}"#, escape(path)),
        ),
    }
}

// ---------- endpoints ----------

fn is_bad_scenario(s: &str) -> bool {
    s == "secret_drift" || s == "breaking_change" || s == "schema_mismatch"
}

/// Legacy: aplica la version sin preflight y deja el ambiente como quede.
fn deploy_legacy(env: &str, version: &str, scenario: &str) -> String {
    LEGACY_DEPLOYS.fetch_add(1, Ordering::Relaxed);

    let outcome = if is_bad_scenario(scenario) {
        LEGACY_BROKEN.fetch_add(1, Ordering::Relaxed);
        DeployOutcome::DeployedButBroken
    } else {
        DeployOutcome::Deployed
    };
    let health = match outcome {
        DeployOutcome::DeployedButBroken => "degraded",
        _ => "healthy",
    };

    let mut guard = STATE.lock().unwrap();
    let state = guard.as_mut().expect("state");
    if let Some(e) = state.environs.iter_mut().find(|e| e.name == env) {
        e.version = version.to_string();
        e.health = health.to_string();
    }
    record(state, "legacy", env, version, scenario, &outcome.label());
    drop(guard);

    format!(
        r#"{{"variant":"legacy","env":"{}","version":"{}","scenario":"{}","result":"{}","health":"{health}","note":"sin preflight ni rollback; ambiente queda como quede."}}"#,
        escape(env),
        escape(version),
        escape(scenario),
        outcome.label()
    )
}

/// Controlled: toda la secuencia bajo un solo lock — leer version, decidir y
/// escribir es una sola transaccion logica.
fn deploy_controlled(env: &str, version: &str, scenario: &str) -> String {
    CONTROLLED_DEPLOYS.fetch_add(1, Ordering::Relaxed);

    let mut guard = STATE.lock().unwrap();
    let state = guard.as_mut().expect("state");
    let before = state
        .environs
        .iter()
        .find(|e| e.name == env)
        .map(|e| e.version.clone())
        .unwrap_or_else(|| "unknown".into());

    let outcome = if scenario == "missing_artifact" || scenario == "secret_drift_detected" {
        CONTROLLED_BLOCKED.fetch_add(1, Ordering::Relaxed);
        DeployOutcome::BlockedInPreflight { current_version: before.clone() }
    } else if is_bad_scenario(scenario) {
        CONTROLLED_ROLLBACKS.fetch_add(1, Ordering::Relaxed);
        DeployOutcome::RolledBack { to_version: before.clone() }
    } else {
        if let Some(e) = state.environs.iter_mut().find(|e| e.name == env) {
            e.version = version.to_string();
            e.health = "healthy".to_string();
        }
        DeployOutcome::Promoted
    };

    record(state, "controlled", env, version, scenario, &outcome.label());
    drop(guard);

    // match exhaustivo: agregar una variante al enum rompe aca la compilacion.
    let body = match &outcome {
        DeployOutcome::BlockedInPreflight { current_version } => format!(
            r#""result":"blocked_in_preflight","current_version":"{}","note":"preflight bloqueo antes de tocar el ambiente.""#,
            escape(current_version)
        ),
        DeployOutcome::RolledBack { to_version } => format!(
            r#""result":"rolled_back","current_version":"{}","note":"smoke fallo, rollback automatico a la version anterior.""#,
            escape(to_version)
        ),
        DeployOutcome::Promoted => {
            r#""result":"promoted","health":"healthy","note":"preflight ok + smoke ok → promote.""#.to_string()
        }
        DeployOutcome::Deployed | DeployOutcome::DeployedButBroken => {
            r#""result":"unexpected","note":"variante legacy en ruta controlada.""#.to_string()
        }
    };

    format!(
        r#"{{"variant":"controlled","env":"{}","version":"{}","scenario":"{}",{body}}}"#,
        escape(env),
        escape(version),
        escape(scenario)
    )
}

fn record(state: &mut State, variant: &str, env: &str, version: &str, scenario: &str, result: &str) {
    state.deployments.insert(
        0,
        Deployment {
            at: rfc3339_now(),
            variant: variant.into(),
            env: env.into(),
            version: version.into(),
            scenario: scenario.into(),
            result: result.into(),
        },
    );
    state.deployments.truncate(MAX_DEPLOYMENTS);
}

fn environments_json() -> String {
    let guard = STATE.lock().unwrap();
    let state = guard.as_ref().expect("state");
    let envs: Vec<String> = state
        .environs
        .iter()
        .map(|e| {
            format!(
                r#"{{"name":"{}","version":"{}","health":"{}"}}"#,
                escape(&e.name),
                escape(&e.version),
                escape(&e.health)
            )
        })
        .collect();
    format!(r#"{{"envs":[{}]}}"#, envs.join(","))
}

fn deployments_json() -> String {
    let guard = STATE.lock().unwrap();
    let state = guard.as_ref().expect("state");
    let history: Vec<String> = state
        .deployments
        .iter()
        .map(|d| {
            format!(
                r#"{{"at":"{}","variant":"{}","env":"{}","version":"{}","scenario":"{}","result":"{}"}}"#,
                escape(&d.at),
                escape(&d.variant),
                escape(&d.env),
                escape(&d.version),
                escape(&d.scenario),
                escape(&d.result)
            )
        })
        .collect();
    format!(
        r#"{{"history":[{}],"max_kept":{MAX_DEPLOYMENTS}}}"#,
        history.join(",")
    )
}

fn diagnostics() -> String {
    format!(
        r#"{{"stack":"{}","case":"{CASE_NAME}","legacy":{{"deploys":{},"broken_state_left":{},"behavior":"sin preflight, sin rollback"}},"controlled":{{"deploys":{},"blocked_in_preflight":{},"rollbacks":{},"behavior":"preflight + smoke + rollback automatico"}},"environments":{}}}"#,
        stack(),
        LEGACY_DEPLOYS.load(Ordering::Relaxed),
        LEGACY_BROKEN.load(Ordering::Relaxed),
        CONTROLLED_DEPLOYS.load(Ordering::Relaxed),
        CONTROLLED_BLOCKED.load(Ordering::Relaxed),
        CONTROLLED_ROLLBACKS.load(Ordering::Relaxed),
        environments_json()
    )
}

// ---------- helpers ----------

fn rfc3339_now() -> String {
    let d = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default();
    format!("{}.{:09}Z", epoch_to_iso(d.as_secs()), d.subsec_nanos())
}

fn epoch_to_iso(secs: u64) -> String {
    let days = secs / 86400;
    let rem = secs % 86400;
    let (h, m, s) = (rem / 3600, (rem % 3600) / 60, rem % 60);
    let mut year = 1970u64;
    let mut d = days;
    loop {
        let leap = (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
        let len = if leap { 366 } else { 365 };
        if d < len {
            break;
        }
        d -= len;
        year += 1;
    }
    let leap = (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
    let months = [31, if leap { 29 } else { 28 }, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    let mut month = 1;
    for len in months {
        if d < len {
            break;
        }
        d -= len;
        month += 1;
    }
    format!("{year:04}-{month:02}-{:02}T{h:02}:{m:02}:{s:02}", d + 1)
}

fn escape(v: &str) -> String {
    v.replace('\\', "\\\\").replace('"', "\\\"")
}
